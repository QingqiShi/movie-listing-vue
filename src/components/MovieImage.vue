<template>
  <div class="movie-image">
    <img :src="srcUrl" :srcset="srcset" :sizes="sizes" alt="" />
  </div>
</template>

<script>
import { mapState } from "vuex";
import api from "@/constants/api";

const { IMAGE, ORIGINAL_IMAGE_WIDTH } = api;

export default {
  props: {
    src: { type: String, default: "" },
    type: { type: String, default: "poster" },
    targetWidth: { type: Number, default: 500 }
  },
  data: () => ({
    imageWidth: 100
  }),
  computed: {
    ...mapState({
      baseUrl: state => state.movies.configuration.imageBaseUrl,
      sizeArray(state) {
        if (!state.movies.configuration[`${this.type}Sizes`]) return [];

        return state.movies.configuration[`${this.type}Sizes`].filter(size =>
          // Numbers leading with 'w' or the word 'original'
          size.match(/^w[0-9]+$|^original$/)
        );
      }
    }),
    /** @returns {string} */
    srcUrl() {
      const size = this.sizeArray.length ? this.sizeArray[0] : "";
      return IMAGE(this.baseUrl, size, this.src);
    },
    srcset() {
      return this.sizeArray
        .map(size => {
          const query =
            size === "original"
              ? `${ORIGINAL_IMAGE_WIDTH}w`
              : `${parseInt(size.slice(1))}w`;

          return `${IMAGE(this.baseUrl, size, this.src)} ${query}`;
        })
        .join(", ");
    },
    sizes() {
      const getSizeQuery = i =>
        `(min-width: ${this.targetWidth * i}px) ${(100 / i).toFixed(2)}vw`;
      const result = [];

      // Loop from 7, above 7 the size changes are minimal
      for (let i = 7; i > 1; i--) {
        result.push(getSizeQuery(i));
      }
      result.push("100vw");

      return result.join(", ");
    }
  }
};
</script>

<style lang="scss">
.movie-image {
  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
}
</style>
