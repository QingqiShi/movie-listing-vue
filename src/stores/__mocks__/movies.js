const state = {
  configuration: {
    imageBaseUrl: "https://images.com/",
    posterSizes: ["w100", "w300", "original"],
    backdropSizes: ["w100", "w300", "original"]
  },
  movies: [
    {
      id: 1000,
      title: "Avatar",
      genre_ids: [0],
      popularity: 230,
      vote_average: 5
    }
  ],
  genres: [{ id: 0, name: "Action" }, { id: 1, name: "Horror" }],
  sortOrder: false,
  sortProperty: "popularity",
  filterRating: 3,
  filterGenres: { 0: false, 1: true },
  error: ""
};

const getters = {
  /** @param {typeof state} state */
  ratingFilteredMovies: state => state.movies,
  availableGenres: () => [{ id: 0, name: "Action" }],
  filteredMovies: () => []
};

const mutations = {
  setConfiguration: jest.fn(),
  setMovies: jest.fn(),
  setGenres: jest.fn(),
  setSort: jest.fn(),
  setFilter: jest.fn(),
  setError: jest.fn()
};

const actions = {
  initialFetch: jest.fn(() => Promise.resolve()),
  fetchConfiguration: jest.fn(() => Promise.resolve()),
  fetchMovies: jest.fn(() => Promise.resolve()),
  fetchGenres: jest.fn(() => Promise.resolve())
};

export default { state, getters, mutations, actions };
