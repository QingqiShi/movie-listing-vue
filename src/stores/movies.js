import api from "@/constants/api";
import app from "@/constants/app";

const { CONFIGURATION, NOW_PLAYING, GENRES_LIST } = api;
const { LOCALE } = app;

const state = {
  configuration: {},
  /** @type {Movie[]} List of all movies */
  movies: [],
  /** @type {Genre[]} List of all genres */
  genres: [],
  /** @type {boolean} True for accending, false for decending */
  sortOrder: false,
  /** @type {"popularity"|"vote_average"} */
  sortProperty: "popularity",
  /** @type {number} Number between 0 and 10 */
  filterRating: 3,
  /** @type {{[id:number]: boolean}} */
  filterGenres: {},
  /** @type {string} */
  error: ""
};

const getters = {
  /** @type {(state:State) => Movie[]} */
  ratingFilteredMovies: state =>
    state.movies.filter(movie => movie.vote_average >= state.filterRating),

  /** @type {(state:State, getters:Getters) => Genre[]} */
  availableGenres: (state, getters) =>
    state.genres.filter(genre => {
      for (let movie of getters.ratingFilteredMovies) {
        if (movie.genre_ids.includes(genre.id)) {
          return true;
        }
      }
      return state.filterGenres[genre.id];
    }),

  /** @type {(state:State, getters:Getters) => Movie[]} */
  filteredMovies: (state, getters) => {
    const { availableGenres, ratingFilteredMovies } = getters;
    const { filterGenres } = state;

    const genres = availableGenres.filter(genre => filterGenres[genre.id]);
    if (!genres.length) return ratingFilteredMovies;

    return ratingFilteredMovies.filter(({ genre_ids }) =>
      genres.reduce((show, genre) => show && genre_ids.includes(genre.id), true)
    );
  }
};

const mutations = {
  /** @type {(state:State, payload:Configuration) => void} */
  setConfiguration(state, { imageBaseUrl, posterSizes, backdropSizes }) {
    state.configuration = { imageBaseUrl, posterSizes, backdropSizes };
  },

  /** @type {(state:State, movies:Movie[]) => void} */
  setMovies(state, movies) {
    state.movies = movies;
  },

  /** @type {(state:State, genres:Genre[]) => void} */
  setGenres(state, genres) {
    state.genres = genres;
    state.filterGenres = {};
    genres.forEach(genre => {
      state.filterGenres[genre.id] = false;
    });
  },

  /** @type {(state:State, payload:{property:SortProperty, order:boolean}) => void} */
  setSort(state, { property, order }) {
    state.sortOrder = order;
    state.sortProperty = property;
    state.movies = state.movies.sort((a, b) =>
      order ? a[property] - b[property] : b[property] - a[property]
    );
  },

  /** @type {(state:State, payload:{type:"rating"|"genres", value:number}) => void} */
  setFilter(state, { type, value }) {
    if (type === "rating") {
      state.filterRating = value;
    } else {
      const newFilters = { ...state.filterGenres };
      newFilters[value] = !state.filterGenres[value];
      state.filterGenres = newFilters;
    }
  },

  /** @type {(state:State, errorMsg:string) => void} */
  setError(state, errorMsg) {
    state.error = errorMsg;
  }
};

const actions = {
  /** @param {import('vuex').ActionContext} context */
  async initialFetch({ commit, dispatch }) {
    try {
      await Promise.all([
        dispatch("fetchConfiguration"),
        dispatch("fetchMovies"),
        dispatch("fetchGenres")
      ]);
      commit("setSort", { property: "popularity", order: false });
    } catch (e) {
      commit("setError", e);
    }
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchConfiguration({ commit }) {
    const result = await fetchApi(
      CONFIGURATION(),
      "FetchConfiguration",
      result => result && result.images
    );

    commit("setConfiguration", {
      imageBaseUrl: result.images.base_url,
      posterSizes: result.images.poster_sizes,
      backdropSizes: result.images.backdrop_sizes
    });
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchMovies({ commit }) {
    const result = await fetchApi(
      NOW_PLAYING(LOCALE),
      "FetchMovies",
      result => result && result.results
    );

    commit("setMovies", result.results);
  },

  /** @param {import('vuex').ActionContext} context */
  async fetchGenres({ commit }) {
    const result = await fetchApi(
      GENRES_LIST(LOCALE),
      "FetchGenres",
      result => result && result.genres
    );

    commit("setGenres", result.genres);
  }
};

/**
 * @param {string} url
 * @param {string} name
 */
async function fetchApi(url, name, validator) {
  let result;

  try {
    const response = await fetch(url);
    result = await response.json();
  } catch {
    throw `${name} failed, check your network and try again`;
  }

  if (!validator(result)) {
    throw `${name} returned invalid result`;
  }

  return result;
}

export default { state, getters, mutations, actions };

/* Type definitions to help vscode do auto suggest */

/** @typedef {{id:number, name:string}} Genre */
/**
 * @typedef {object} Movie
 * @property {number} Movie.id
 * @property {string} Movie.title
 * @property {number} Movie.popularity
 * @property {number} Movie.vote_average
 * @property {number[]} Movie.genre_ids
 * @property {string} Movie.poster_path
 */
/**
 * @typedef {object} Configuration
 * @property {string} Configuration.imageBaseUrl
 * @property {string} Configuration.posterSizes
 * @property {string} Configuration.backdropSizes
 */
/** @typedef {typeof state} State */
/**
 * @typedef {object} Getters
 * @property {Movie[]} Getters.ratingFilteredMovies
 * @property {Genre[]} Getters.availableGenres
 * @property {Movie[]} Getters.filteredMovies
 */
/** @typedef {typeof state.sortProperty} SortProperty */
