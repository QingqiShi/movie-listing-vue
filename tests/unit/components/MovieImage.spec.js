import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";
import MovieImage from "@/components/MovieImage.vue";
import movies from "@/stores/movies";

jest.mock("@/stores/movies");

const localVue = createLocalVue();
localVue.use(Vuex);

describe("MovieImage.vue", () => {
  let store;

  beforeAll(() => {
    store = new Vuex.Store({ modules: { movies } });
  });

  it("renders", () => {
    const wrapper = shallowMount(MovieImage, {
      store,
      localVue,
      propsData: { src: "/test.jpg", type: "poster", targetWidth: 300 }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("invalid type", () => {
    const wrapper = shallowMount(MovieImage, {
      store,
      localVue,
      propsData: { src: "/test.jpg", type: "invalid", targetWidth: 300 }
    });
    expect(wrapper).toMatchSnapshot();
  });
});
