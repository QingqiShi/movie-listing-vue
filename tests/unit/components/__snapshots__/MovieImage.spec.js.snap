// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieImage.vue invalid type 1`] = `<div class="movie-image"><img src="https://images.com//test.jpg" srcset="" sizes="(min-width: 2100px) 14.29vw, (min-width: 1800px) 16.67vw, (min-width: 1500px) 20.00vw, (min-width: 1200px) 25.00vw, (min-width: 900px) 33.33vw, (min-width: 600px) 50.00vw, 100vw" alt=""></div>`;

exports[`MovieImage.vue renders 1`] = `<div class="movie-image"><img src="https://images.com/w100/test.jpg" srcset="https://images.com/w100/test.jpg 100w, https://images.com/w300/test.jpg 300w, https://images.com/original/test.jpg 2000w" sizes="(min-width: 2100px) 14.29vw, (min-width: 1800px) 16.67vw, (min-width: 1500px) 20.00vw, (min-width: 1200px) 25.00vw, (min-width: 900px) 33.33vw, (min-width: 600px) 50.00vw, 100vw" alt=""></div>`;
